### DISCUSSION TITLE
Grasping the nettle.  Epic replacement of ShuttleGui

### DISCUSSION THREAD

**Discussion by Paul-Licameli** [audacity/audacity#5628] [Development]
# What is it?

`ShuttleGui` is a widely used library in Audacity that simplifies the construction of dialogs, using simplified nested constructions resembling the logical nesting of dialog controls in panels, and hiding many lower level calls to wxWidgets that build up the dialogs.  It's somewhat like `qml` but defined in C++ and invented years before that.

And yet, the system has many shortcomings and is now seen as an inhibition to the "greening" of many parts of the program, because it is used in a great number of places and is too closely tied to wxWidgets.

But don't give up hope -- I think it is possible to write a library intelligently so that the routine "transliteration" of most or all of the existing usage of ShuttleGui into alternative new C++ notation is a reasonable goal.

Rather than duplicating all in qml for the developing replacement, which the existing wxWidgets based Audacity couldn't use during the transitional period of we don't know how many releases, I would rather take this approach, allowing much larger parts of the existing sources to be re-used in both Audacity versions unchanged.

Besides which, the "pluggability" property, which certain modules (such as lib-ffmpeg, lib-mp3) would rely on for decoupling from Audacity, is not one that a static source tree of qml files would enjoy.

# What would this un-block?

Let's call the hypothetical library that would define the framework, `lib-dialogs`.  We must keep it "green" -- toolkit-neutral, not dependent on wxWidgets besides the wxBase layer.

Many things in Audacity might then be re-implemented on the new API, cutting their transitive build-time dependencies on wxCore, and also pulled out of the "pentagon" in the modules dependency graph as green libraries.

In the present state of the dependency graph of files inside the "pentagon," several of the pages within the Preferences dialog are near the bottom, and so ripe for extraction into libraries.  We could do that tomorrow, though they would be "red".  With this extra work, they may become "green."

Such work would also be a necessary step toward breaking the dependency of some of the import-export modules on the Audacity executable, and also, turning them "green."

Many existing built-in effects would also require this to turn them "green."

Extraction of `ProjectFileManager` and `ProjectManager` from the pentagon into libraries, and several of the `src/menus` files implementing pull-down menu item actions, would also require this.

# What would the new notation look like?

I'm still working that out, but with some operator overloading for a "named parameter" idiom, it might look like this.  To take a small example from `ImportStreamDialog.cpp`:

```
   ShuttleGui S{ this, eIsCreating };
   {
      S.SetBorder( 5 );

      StreamList =
      S
         .Prop(1)
         .Position(wxEXPAND | wxALIGN_LEFT | wxALL)
         .Style(wxLB_EXTENDED | wxLB_ALWAYS_SB)
         .AddListBox(
            transform_container<wxArrayStringEx>(
               mFile->GetStreamInfo(),
               std::mem_fn( &TranslatableString::Translation ) ) );

      S.AddStandardButtons();
   }

...
   
void ImportStreamDialog::OnOk(wxCommandEvent & WXUNUSED(event))
{
   wxArrayInt selitems;
   int sels = StreamList->GetSelections(selitems);
   for (wxInt32 i = 0; i < sels; i++)
      mFile->SetStreamUsage(selitems[i],TRUE);
   EndModal( wxID_OK );
}

void ImportStreamDialog::OnCancel(wxCommandEvent & WXUNUSED(event))
{
   EndModal( wxID_CANCEL );
}
```

Instead one might write something like:

```
std::vector<size_t>  selitems;

using namespace Dialogs;
static const auto dialog = (Vertical / Border(5)) (
   ListBox{ mFile->GetStreamInfo() }
      - Prop(1),
      - Position(Expand, Left, All)
         >> selitems, // associating control with a variable
   StandardButtons);
   
ShowModal(dialog);

for (auto item : selitems)
   mFile->SetStreamUsage(item, true);
```

Notice that `wxEvent` handler functions are eliminated.

# The five properties of open-endedness

The ideal system should be "open ended" in five different ways, meaning that extensions along any one of these five "axes" should be achievable in separately compiled files, once an adequate basic library is defined, requiring no more intrusions into that library; but also meaning that particular "declaratively" specified expression trees (see the first point), defined using this library, can also be extended non-intrusively with yet other separately compiled files (the second point).

## Declarativity, or, variation of visitation actions

All the present uses of ShuttleGui are procedural -- making a sequence of calls on a `ShuttleGui` object, usually given the variable name `S` -- and `enum teShuttleMode` is the type of a member variable of that object.

But instead, the ideal system should be declarative, like the utilities in the recent `lib-menus`.  That means, to use it one writes what is basically a long, and nested, but single constructor expression which returns an object that describes the tree-structure of a dialog.  (Non-procedurality still allows some embedded lambda expressions, describing delayed evaluations or executions, such as, what a button press does.)

Then, to use this object, one "visits" it, and the visitation actions should be easy to vary to accomplish different things -- most often, building a dialog, but sometimes too, just collecting information from the tree, as in `GetInfoCommand.cpp` (eliminating need for the class `ShuttleGuiGetDefinition`)

## Pluggability, or, reuse the Registry system

An existing usage of ShuttleGui is "closed ended."   The associated dialog can't be modified, except intrusively in that particular usage of ShuttleGui.

There is an exception to this now in `LibraryPrefs.cpp` introduced at commit 6bd5b0f66c3fcf1898db7de29c0b0f52ae9b2677 -- an ad-hoc registry of functions that are given the `ShuttleGui` object, to insert more controls.  (This was a dependency-breaking change that enabled the extraction of import/export modules for MP3 and FFmpeg.)

But more convenient would be a more general utility to make such extensions easier.

If you are familiar with `lib-registries` and `lib-menus`, you know where this leads:  `lib-dialogs` should also use `lib-registry` and make it easy to define the analogue of `AttachedItem` but for dialogs instead of menus.

`lib-registries` also makes it very easy to define ad-hoc "visitors" with a few lambda expressions, addressing the first point.

## Extensibility

To add any new kind of control (basic kinds in any UI library, like button, slider, etc., or even more complicated custom kinds like `NumericTextCtrl`, etc.), or kind of grouping item (Panel, layout sizers of various kinds [vertical, horizontal, grid, wrapping], scroller, static text box, etc.) -- the existing `ShuttleGui.h` and its companion `.cpp` file must be disturbed, defining a new member function of that class.

A better system would use subclassing of the node and leaf types of the registry with virtual functions to do such things as build a control or refresh its appearance; then each kind of control defines its overrides.

## Toolkit-neutrality

`ShuttleGui.h` is littered with mention of many wxWidgets classes.

Intead, each kind of abstract control can define its own tiny factory façade using `DefaultedGlobalHook`.  See for instance `VSTEffectsModule::Factory` and the bottom of `VSTEffect.cpp` for an analogy.

Perhaps a new, red `lib-wx-dialogs` could concentrate most of these injected functions for the common kinds of controls, while some `lib-qt-dialogs` would provide the alternative implementations in Audacity 4.

See also `wxWidgetsBasicUI.cpp` in `lib-wx-init` for another analogy.

## Recombination

By this I mean, an abstract, "generalized variable" base class (with a template type parameter!) could be known in `lib-dialogs`, and the abstract controls would interact with these to move data to the graphical display, or from it in response to user events.

Reimplementations of the generalized variable could shuttle values to simple in-memory variables, or to `Settings` objects in preferences, or more complicated possibilities would include value transformations (such as from linear amplitude to deciBels and back).

Contrast this with, for instance, three different overloads of `TieChoice` in `ShuttleGui.h`.  Requiring a different overload whenever the details of storage of the associated variable changes is unsatisfatory.

# Does it exist yet?

Some of it.  This long branch https://github.com/Paul-Licameli/audacity/tree/tie-functions has been in development a very long time -- some commits with author dates in 2020 or earlier!  I hadn't articulated all of the open-endednesses to myself yet.  I didn't invent nifty notation with operator overloading.  Not all of this is usable, and Qt migration wasn't on my mind then.   I didn't achieve declarativity or pluggability, but surely had imagined them.

But I did go so far as to eliminate `enum teShuttleMode`, I did implement generalized variables, I did make it easy to bind an action to a button or a control with a simple lambda supplied in the ShuttleGui syntax.  I did simplify many of the built-in effects, eliminating the need for many event handler functions, the need to retain pointers to the wxControl objects, the need to include the wxWidgets header files for many control classes in many places.

I did simplify the bi-direction association of, say, a slider and a textbox in a dialog, showing the same value in two different forms, letting the user modify either one and having the other update.

Compare for instance this shortened version of the Amplify effect with what is in master now.  Inclusions of many headers are eliminated, with the event dispatch table, pointers to controls, numerical Ids assigned to them, and the TransferData(From/To)Window functions. https://github.com/Paul-Licameli/audacity/blob/tie-functions/src/effects/Amplify.cpp

# The grand checklist of rewrites

Don't panic!  This looks long and large, but the first few worked out examples will likely be the hardest, after which the others will become routine and repetitious and easier to test.

## Preference pages:
- [ ] BatchPrefs
- [ ] DevicePrefs
- [ ] DirectoriesPrefs
- [ ] EffectsPrefs
- [ ] ExtImportPrefs
- [ ] GUIPrefs
- [ ] ImportExportPrefs
- [ ] KeyConfigPrefs
- [ ] LibraryPrefs
- [ ] MidiIOPrefs
- [ ] ModulePrefs
- [ ] MousePrefs
- [ ] PlaybackPrefs
- [ ] QualityPrefs
- [ ] RecordingPrefs
- [ ] SpectrumPrefs
- [ ] ThemePrefs
- [ ] TracksBehaviorsPrefs
- [ ] TracksPrefs
- [ ] WarningsPrefs
- [ ] WaveformPrefs
- [ ] and PrefsDialog itself which contains the pages!

## Others
- [ ] ErrorDialog
- [ ] HelpSystem::ShowInfoDialog
- [ ] HelpSystem::ShowHtmlText
- [ ] LogWindow::Show
- [ ] MultiDialog
- [ ] AudacityCommandDialog
- [ ] LongMessageDialog
- [ ] VSTEffectOptionsDialog
- [ ] VST3OptionsDialog
- [ ] AudioUnitEffectOptionsDialog
- [ ] LadspaEffectOptionsDialog
- [ ] LV2PreferencesDialog
- [ ] NyquistOutputDialog
- [ ] ContrastDialog
- [ ] EffectPresetsDialog
- [ ] EffectUIHost
- [ ] EffectUIHost
- [ ] EffectUIHost::OnSaveAs
- [ ] EffectDialog
- [ ] EqualizationCurvesDialog
- [ ] ScoreAlignDialog
- [ ] ExportAudioDialog
- [ ] CustomSampleRateDialog
- [ ] ExportFilePanel
- [ ] ExportMixerDialog
- [ ] ImportRawDialog
- [ ] ImportStreamDialog
- [ ] QuickFixDialog
- [ ] OnResample
- [ ] DeviceToolBar
- [ ] LabelTrackMenuTable
- [ ] ChangeClipSpeedDialog
- [ ] RateMenuTable
- [ ] SliderDialog
- [ ] MeterPanel
- [ ] MissingPluginsErrorDialog
- [ ] UnwritableLocationErrorDialog
- [ ] WarningDialog
- [ ] AboutDialog
- [ ] AudacityFileConfig
- [ ] AudioPasteDialog
- [ ] AutoRecoveryDialog
- [ ] MacroCommandDialog
- [ ] ApplyMacroDialog
- [ ] ApplyMacroToProject
- [ ] OnApplyToFiles
- [ ] MacrosWindow
- [ ] BenchmarkDialog
- [ ] FrameStatisticsDialog
- [ ] FrequencyPlotDialog
- [ ] ShowDiagnostics
- [ ] HistoryDialog
- [ ] LabelDialog
- [ ] LangChoiceDialog
- [ ] PluginRegistrationDialog
- [ ] CompactDialog
- [ ] ScreenshotBigDialog
- [ ] SoundActivatedRecordDialog
- [ ] SpectralDataDialog
- [ ] SplashDialog
- [ ] TagsEditorDialog
- [ ] TagsEditorDialog::OnEdit
- [ ] TimeDialog
- [ ] TimerRecordDialog
- [ ] TimerRecordExportDialog
- [ ] FindDialog (in ExportMp3.cpp)
- [ ] MP3ExportProcessor::AskResample
- [ ] CLExportProcessor
- [ ] NyqBench
- [ ] FFmpegExporter
- [ ] ExportFFmpegOptions
- [ ] FindFFmpegDialog
- [ ] FFmpegNotFoundDialog

**Comment by vsverchinsky** [comment]
Pontential roadmap for this epic:
1) Invent new `ShuttleGui` API
2) Implement `ShuttleGui-wx` backend
3) Rewrite all existing UI in Audacity 3.x branch with the new syntax
4) Implement `ShuttleGui-qtquick` backend
5) Manually replace existing `ShuttleGui` code with components written in qml in Audacity 4.x branch (do we still want components to be shared between Audacity and MuseScore?)

Is that correct?

**Reply by Paul-Licameli** [reply]
Number 3 need not be all or nothing.  Selected dialogs could be rewritten where it's most advantageous near term to make a library green.

I don't understand the distinction between 4 and 5.  Explain to me.

**Reply by vsverchinsky** [reply]
`ShuttleGui-qtquick` backend allows execution of existing code written C++/ShuttleGui API in QtQuick environment. The benifit is that all existing UI code written for 3.x will run in 4.x too, "out-of-the-box". But eventually Audacity 4.x will get it's own design based on UI components shared between Audacity and MuseScore.

**Comment by crsib** [comment]
I proposed this approach back in 2021, but few things have been learned since. It is now clear, that there is no easy way to "upgrade" dialogs to the neutral version of "ShuttleGui". This document proves this point well.

Initially, the idea was that we could migrate relatively easily and get a large hunk of Audacity running within Qt. Next, we would polish the UI for each screen step by step. Qt version of Audacity will have a different UI in most, if not all, areas. We do not aim to keep the old UI but within a new UI engine. For this reason, migrating to `NewShuttleGui` seems like an excessive step, honestly.

> ShuttleGui is a widely used library in Audacity that simplifies the construction of dialogs

After 2.5 years in Audacity, doing loads of UI stuff, I have to say that using ShuttleGui makes writing the UI code to be orders of magnitude more complex. The problem with ShuttleGui is not that it is not easily extensible - after all, it has a workaround to host any controls. However, the implementation makes it increasingly hard to write UI even using the existing controls. The same problem applies to few other parts of Audacity.

I must change the ShuttleGui for half of the UI tasks I am assigned. I do not extend it, I try to make it work. I try to do that carefully, but there are always chances to break some distant dialog.  

> allowing much larger parts of the existing sources to be re-used in both Audacity versions unchanged

This document alone lists **95** places that we will have to change in a nontrivial way. Not to mention that we need to implement a very complex system from scratch first.

> Besides which, the "pluggability" property, which certain modules (such as lib-ffmpeg, lib-mp3) would rely on for decoupling from Audacity, is not one that a static source tree of qml files would enjoy

QML has modules, too. There is no "static list of files". They are used right now to define complex components anyway. 

UI in modules is an interesting topic, though. Probably we need to have a concept of "dependent modules" of some kind, so we can have `mod-a` and `wx/mod-a-ui.`

--- 

I would call for another approach, and you use this approach a lot. In fact, you have proposed it and demonstrated how well it works. Instead of trying to fix HOW UI is done in places where UI should NOT BE in the first place - we decouple the UI from the code in question using hooks, callbacks, or observability. This should be evaluated carefully for each scenario. 

For example, for `ProjectFileManager`, the `CompactDialog` should not be there. Instead, it should be shown by the caller. Ironically, `ProjectFileManager::Compact` is not used at all now.

In some cases, like MP3 export, we should drop the code. See in the ExportMP3.
```c++
//This call should go away once export project rate option
//is available as an export dialog option
rate = AskResample(bitrate, rate, lowrate, highrate);
```

Remember yourself saying: our goal is not to figure out how to switch to a different UI toolkit; our goal is to figure out how to build Audacity without UI at all.

**Comment by Paul-Licameli** [comment]
> I proposed this approach back in 2021, but few things have been learned since. It is now clear, that there is no easy way to "upgrade" dialogs to the neutral version of "ShuttleGui". This document proves this point well.
> 
> Initially, the idea was that we could migrate relatively easily and get a large hunk of Audacity running within Qt. Next, we would polish the UI for each screen step by step. Qt version of Audacity will have a different UI in most, if not all, areas. We do not aim to keep the old UI but within a new UI engine. For this reason, migrating to `NewShuttleGui` seems like an excessive step, honestly.
> 
> > ShuttleGui is a widely used library in Audacity that simplifies the construction of dialogs
> 
> After 2.5 years in Audacity, doing loads of UI stuff, I have to say that using ShuttleGui makes writing the UI code to be orders of magnitude more complex. The problem with ShuttleGui is not that it is not easily extensible - after all, it has a workaround to host any controls. However, the implementation makes it increasingly hard to write UI even using the existing controls. The same problem applies to few other parts of Audacity.
> 

I do not like ShuttleGui.  I think it could be a better abstraction layer in all the ways I detailed.  But this old rationale from James describes how it improved on what preceded it.  https://aosabook.org/en/v1/audacity.html

> I must change the ShuttleGui for half of the UI tasks I am assigned. I do not extend it, I try to make it work. I try to do that carefully, but there are always chances to break some distant dialog.
> 

I must look through commit history to see what you mean.  Was ShuttleGui or wxWidgets the real source of difficulty?

> > allowing much larger parts of the existing sources to be re-used in both Audacity versions unchanged
> 
> This document alone lists **95** places that we will have to change in a nontrivial way. Not to mention that we need to implement a very complex system from scratch first.

As I said, replacement need not be all or none.  Still it could be attempted in a topic branch. A few small successful examples might prove the concept.

> 
> > Besides which, the "pluggability" property, which certain modules (such as lib-ffmpeg, lib-mp3) would rely on for decoupling from Audacity, is not one that a static source tree of qml files would enjoy
> 
> QML has modules, too. There is no "static list of files". They are used right now to define complex components anyway.

Can they do what the Registry system does for menus, let’s say?

Did I waste time developing registries?
 
> UI in modules is an interesting topic, though. Probably we need to have a concept of "dependent modules" of some kind, so we can have `mod-a` and `wx/mod-a-ui.`
> 
> I would call for another approach, and you use this approach a lot. In fact, you have proposed it and demonstrated how well it works. Instead of trying to fix HOW UI is done in places where UI should NOT BE in the first place - we decouple the UI from the code in question using hooks, callbacks, or observability. This should be evaluated carefully for each scenario.

These are not mutually exclusive ideas.

At some high enough level, there must be some interaction with the user.

But still why not develop a toolkit neutral vocabulary describing logical dialog structure and the association of controls with variables?  Then even that high level might be shareable source code between the programs.

> 
> For example, for `ProjectFileManager`, the `CompactDialog` should not be there. Instead, it should be shown by the caller. Ironically, `ProjectFileManager::Compact` is not used at all now.
> 
> In some cases, like MP3 export, we should drop the code. See in the ExportMP3.
> 
> ```c++
> //This call should go away once export project rate option
> //is available as an export dialog option
> rate = AskResample(bitrate, rate, lowrate, highrate);
> ```
> 
> Remember yourself saying: our goal is not to figure out how to switch to a different UI toolkit; our goal is to figure out how to build Audacity without UI at all.

I have not exactly said that.  I have said we should avoid static build dependency on this or that UI toolkit.  But sometimes the expedient to migrate the code we have is not to eliminate all UI code, but instead use facades as with BasicUI.  The new `namespace Dialogs` would be a more elaborate facade for less basic UI.

**Reply by crsib** [reply]
> Was ShuttleGui or wxWidgets the real source of difficulty?

It was 100% Shuttle, and its little implementation details. 

> Can they do what the Registry system does for menus, let’s say?

I do not quite understand the unique properties of Registry and I don't follow how is it relevant to UI system. For example, NumericTextContol is now a plugin with a separate source tree.

> Still it could be attempted in a topic branch. A few small successful examples might prove the concept.

I expect it to take months to develop such a solution. 

> These are not mutually exclusive ideas.

It is common practice to separate UI code from the rest of the application for the past decades. BasicUI helps to quickly make libraries independent, but it encourages keeping bad practices in code.  

> But sometimes the expedient to migrate the code we have is not to eliminate all UI code

And again, BasicUI is one case, because it largely stands for trivial changes. Rewriting UI using a new, complex abstraction is a different story. It is surely much more difficult than just separating the code using hooks.

**Comment by Paul-Licameli** [comment]
https://aosabook.org/en/v1/audacity.html

Amusing reading now after many years.

Ouch!  The GetLink Story!  I cleaned up those hundred places, inserting the iterations over TrackList::Channels and mostly generalizing away from the assumption of at most two channels.

It “improved in what preceded it” when I did it.  But now, look at me. :😞

**Reply by Paul-Licameli** [reply]
And who made TrackPanel less “horrible”?

Yeah.

**Comment by Paul-Licameli** [comment]
> > Was ShuttleGui or wxWidgets the real source of difficulty?
> 
> It was 100% Shuttle, and its little implementation details.

So if the difficulty of extensibility was the fault of existing ShuttleGui but not so much wxWidgets, this does not argue that a smarter replacement can't be more conveniently extensible.

> 
> > Can they do what the Registry system does for menus, let’s say?
> 
> I do not quite understand the unique properties of Registry and I don't follow how is it relevant to UI system. For example, NumericTextContol is now a plugin with a separate source tree.
> 

There are other uses of Registry.  The most important, and original, one was the declarative menu system we have, and you reviewed extract-lib-menus only recently.

I refer to this.  Is all of that, enabling scattered `AttachedItems`, still worthwhile in Audacity 4 or are you saying it's only wheel-reinvention?

> > Still it could be attempted in a topic branch. A few small successful examples might prove the concept.
> 
> I expect it to take months to develop such a solution.

Months of thought and experiment (long ago) has preceded this already.

> 
> > These are not mutually exclusive ideas.
> 
> It is common practice to separate UI code from the rest of the application for the past decades. BasicUI helps to quickly make libraries independent, but it encourages keeping bad practices in code.
> 
> > But sometimes the expedient to migrate the code we have is not to eliminate all UI code
> 
> And again, BasicUI is one case, because it largely stands for trivial changes. Rewriting UI using a new, complex abstraction is a different story. It is surely much more difficult than just separating the code using hooks.

Is `ExportAudioDialog` from the recent big import-export refactoring a good example of application of this principle?

But look at it:  it's still a non-trivial specification of a dialog layout, using ShuttleGui to get its work done.

It's one obstacle to making FileMenus.cpp tool-kit neutral.

So as I said:  these ideas are not mutually exclusive.  Refactoring the UI properly still leaves us with ShuttleGui and needing an alternative.

So I still think the experiment is work an attempt, and maybe figuring this dialog out first would  be convincing.

**Reply by crsib** [reply]
> So if difficulty of extensibility was the fault of existing ShuttleGui but not so much wxWidgets

There was no difficulty **extending** the ShuttleGui; the problem was **using** it. Extending is not the only concern of software development.

> I refer to this. Is all of that, enabling scattered AttachedItems, still worthwhile in Audacity 4 or are you saying it's only wheel-reinvention?

I don't follow here, sorry. How defining menus/commands is relevant to the UI?

> Is ExportAudioDialog from the recent big import-export refactoring a good example of application of this principle?

No. There are no plans to commit to keeping UI layouts between wxWidgets and Qt in the long term. It would be good if we had such a system two years ago.

> It's one obstacle to making FileMenus.cpp tool-kit neutral.

This is confusing. Why could it not be injected into the registry from the red code? Why does it need to be in a green library?

> Months of thought and experiment (long ago) has preceded this already.

The `tie` branch wasn't quite there when I last looked. In any case, the new library must first find a common denominator between wxWidgets and QML. 

> So I still think the experiment is work an attempt, and maybe figuring this dialog out first would be convincing.

From my experience, this is a very complex task that will require a lot of effort. Keep in mind that writing only `wx` implementation is not sufficient in this case as a proof of concept.

**Comment by Paul-Licameli** [comment]
> > So if difficulty of extensibility was the fault of existing ShuttleGui but not so much wxWidgets
> 
> There was no difficulty **extending** the ShuttleGui; the problem was **using** it. Extending is not the only concern of software development.

Just subjectively, I find ShuttleGui not very difficult to use for simple dialogs.  So I don't know where you found your difficulties or if you recall.

But there is at least one bad thing in its usage -- it is possible to mis-pair various "begin" and "end" commands, and a better declarative notation can eliminate that.

> 
> > I refer to this. Is all of that, enabling scattered AttachedItems, still worthwhile in Audacity 4 or are you saying it's only wheel-reinvention?
> 
> I don't follow here, sorry. How defining menus/commands is relevant to the UI?

Are not menus part of UI too, and something you might do in QML instead?  Will the menu system have lasting value?

> 
> > Is ExportAudioDialog from the recent big import-export refactoring a good example of application of this principle?
> 
> No. There are no plans to commit to keeping UI layouts between wxWidgets and Qt in the long term. It would be good if we had such a system two years ago.
> 
> > It's one obstacle to making FileMenus.cpp tool-kit neutral.
> 
> This is confusing. Why could it not be injected into the registry from the red code? Why does it need to be in a green library?
> 
> > Months of thought and experiment (long ago) has preceded this already.
> 
> The `tie` branch wasn't quite there when I last looked. In any case, the new library must first find a common denominator between wxWidgets and QML.
> 
> > So I still think the experiment is work an attempt, and maybe figuring this dialog out first would be convincing.
> 
> From my experience, this is a very complex task that will require a lot of effort. Keep in mind that writing only `wx` implementation is not sufficient in this case as a proof of concept.

**Reply by crsib** [reply]
> I find ShuttleGui not very difficult to use for simple dialogs

Simple dialogs should be ... simple? Even "not very difficult" is not something desired.

> So I don't know where you found your difficulties or if you recall

Well, I don't always do simple dialogs. ShuttleGui doesn't play well with margins, like at all. Especially when you need a complex structure with nested layouts. Shuttle tends to reset the margins at random points. And the "modern" design is all about margins. Even simple dialogs become a real challenge when you need them to look good (and to have margins).

> Will the menu system have lasting value?

I expect it will.

>  it is possible to mis-pair various "begin" and "end"

That rarely matters. `End*` just calls `PopSizer()` mostly.

**Comment by Paul-Licameli** [comment]
The way `ShuttleGui` handles margins is truly weird and hard to single-step and understand.

This by itself should not detract from the possibility of writing something better.

**Reply by Paul-Licameli** [reply]
It should be easy to express the logical nested structure of a dialog and the association of variables with controls.  (No End... calls but simply parenthetical nesting.)

**Comment by Paul-Licameli** [comment]
> > Will the menu system have lasting value?
> 
> I expect it will.
> 

I'm pleased you think so.  Why is that case different?
